internal/deviations/deviations.go

```diff
@@ -65,27 +65,91 @@ package deviations
+		return fmt.Errorf("vendor should be specified in textproto %v", platformExceptions)
+	}
+	return nil
+}
+
+func lookupDeviations(dut *ondatra.DUTDevice) (*mpb.Metadata_PlatformExceptions, error) {
+	matchedPlatformException := &mpb.Metadata_PlatformExceptions{}
+	isPlatformExceptionsMatched := false
+
 	for _, platformExceptions := range metadata.Get().PlatformExceptions {
+		if err := isPlatformExceptionsValid(platformExceptions); err != nil {
+			return nil, err
+		}
 		if dut.Device.Vendor().String() != platformExceptions.GetPlatform().Vendor.String() {
 			continue
 		}
+		// Exit test if both hardware_model_regex and hardware_model fields are set.
+		if hardwareModelRegex := platformExceptions.GetPlatform().GetHardwareModelRegex(); hardwareModelRegex != "" {
+			// If hardware_model_regex is set and does not match, continue
+			if !matchRegex(hardwareModelRegex, dut.Device.Model()) {
+				continue
+			}
+		}
+		if softwareVersionRegex := platformExceptions.GetPlatform().GetSoftwareVersionRegex(); softwareVersionRegex != "" {
+			// If software_version_regex is set and does not match, continue
+			if !matchRegex(softwareVersionRegex, dut.Device.Version()) {
+				continue
+			}
+		}
+
+		// TODO(prinikasn): Refactor after hardware_model field is removed.
+		if len(platformExceptions.GetPlatform().GetHardwareModel()) == 0 {
+			if isPlatformExceptionsMatched {
+				return nil, fmt.Errorf("cannot have more than one match within platform_exceptions fields %v and %v", matchedPlatformException, platformExceptions)
+			}
+			matchedPlatformException = platformExceptions
+			isPlatformExceptionsMatched = true
+		}
+		// TODO(prinikasn): Remove after hardware_model field is removed.
+		// If hardware_model_regex is empty, get the deviations from matching the model in the hardware_model field.
 		for _, hardwareModel := range platformExceptions.GetPlatform().HardwareModel {
 			if dut.Device.Model() == hardwareModel {
-				return platformExceptions.GetDeviations()
+				if isPlatformExceptionsMatched {
+					return nil, fmt.Errorf("cannot have more than one match within platform_exceptions fields %v and %v", matchedPlatformException, platformExceptions)
+				}
+				matchedPlatformException = platformExceptions
+				isPlatformExceptionsMatched = true
 			}
 		}
 	}
-	log.Warningf("No platform exceptions for dut platform %v or model %v configured in platform exceptions metadata %v", dut.Device.Vendor().String(), dut.Device.Model(), metadata.Get().PlatformExceptions)
-	return &mpb.Metadata_Deviations{}
+	return matchedPlatformException, nil
+}
+
+func lookupDUTDeviations(dut *ondatra.DUTDevice) *mpb.Metadata_Deviations {
+	platformExceptions, err := lookupDeviations(dut)
+	if err != nil {
+		log.Exitf("Error looking up deviations: %v", err)
+	}
+	if platformExceptions.GetPlatform().Vendor.String() == "" {
+		log.Errorf("Did not match any platform_exception %v, returning default values", metadata.Get().GetPlatformExceptions())
```

Comment:
This should only be logged at an Info level. There is no error if there are no deviations -- that's what we ideally want :)

Reply:
Agree. done